Add Wayland driver for devdraw

eaburns opened the pull request.

eaburns committed:
Add wayland support to devdraw.

eaburns committed:
Link the snarf buffer to the compositor's copy/paste buffer on Wayland.

ylh committed:
devdraw/wayland.c: prevent divide-by-zero in scroll repeat

droyo committed:
Fix race condition on attach and mouse track on warp

I tried running the wayland devdraw implementation on Sway 1.10,
wlroots 0.18 and encountered the error:

	xdg_surface#13: error 3: xdg_surface has never been configured

According to https://wayland.app/protocols/xdg-shell#xdg_surface ,
a client must commit a surface without a buffer, *wait* for the first
configure request from the compositor, ack it, and *only then* can
it proceed to attach a buffer to the surface and tell wayland to display it.

This was caused by the following sequence of events:

1. devdraw starts, enters gfx_main
2. `gfx_main` calls `gfx_started`, which spawns the `serveproc` thread
3. `gfx_main` enters `wl_display_dispatch`, flushing any buffered requests to
   the compositor, and enters `wl_display_poll()` to wait for incoming messages
4. `serveproc` calls `rpc_attach`, sets up the surface, and buffers a commit.

The race is between #3 and #4. If #3 happens first, the buffered commit
just sits there until `rpc_flush` is called, which calls `wl_display_flush()`,
but at that point a buffer is attached too quickly for the configure to happen.

This commit fixes the race by adding a `configured` field to the WaylandClient
and using it to guard `rpc_flush`.

In addition, I found that mouse warping, at least in sway, would move
the cursor but future mouse presses would register at the old location
until I moved the mouse. So I added a call to gfx_mousetrack to the end
of `rpc_setmouse`.

droyo committed:
Fix unexpected devdraw crashes on sway

I found sway 1.10 and wlroots 0.18.2 will occasionally generate
wl_keyboard::keymap events with an empty keymap. I do not have
a reliable reproduction for this, but it seems to occur when
switching back and forth between workspaces with the keyboard.

This also fixes the issue where devdraw can go into a busy loop if
the compositor goes away and wl_display_dispatch() returns -1.

droyo committed:
Only register repeat callbacks when repeat active

This change splits scroll repeat and key repeat into separate
callbacks, and only registers them when a key is pressed or
scrolling is active, respectively. The key repeat logic is
changed to support key repetitions higher than the frame rate.

The new logic will also accomodate key repeat parameters set by
the compositor, although the comment about us not receiving that
configuration event is still valid.

droyo committed:
Receive key repeat info from compositor

Upgrade to version 4 of the wl_seat interface which adds the
wl_keyboard::repeat_info event allowing the compositor to configure
key repeat delay and rate.

droyo committed:
move declarations to first use

lneely commented:
I have been using this fork for awhile now, and it works quite well. Recommend merging. :+1:

tusharhero commented:
Works for me. But it checks for ~$DISPLAY~ instead of ~$WAYLAND_DISPLAY~,
#+begin_src bash
  DISPLAY=$WAYLAND_DISPLAY 9 acme
#+end_src

Fixes it, but perhaps you can make it look for ~$WAYLAND_DISPLAY~ directly.

lneely committed:
fix build when WSYSTYPE=wayland in local cfg

lneely committed:
short comment explaining the wayland.c include

JLanzuisi commented:
I've also been testing this branch, great work! 

I've encountered a bug I can't put my finger on: alt-tabbing to a different window and then alt-tabbing back adds spurious characters. I've attached a quick screencast showing the issue.

https://github.com/user-attachments/assets/ac081aa8-fea5-46de-b89c-ace0e51059e1

This is on wayland on gnome 47.

erentar commented:
Thank you! I can finally use acme!

eaburns committed:
Cancel compose when clicking the mouse on an alt+click.

The alt key begins a compose. When using alt+click, compose was not previously cancelled and the next two typed characters would often end up composed.

eaburns committed:
Remove scroll repeat.

This is OK with the touch pad, but it's actually not great with a real mouse scroll wheel, because the scrolling becomes inconsistent and hard to predict exactly how much the screen will scroll. Better to just remove it.

eaburns committed:
Fix cursor background transparency.

It was not transparent, but drawn over a white square.

eaburns committed:
Don't call gfx_abortcompose with a lock held.

There has been some funny locking/stalling behavior that seems to be related to using alt+click as a simulated middle-click. It may have started after this gfx_abortcompose call was added. According to devdraw.h, the gfx_ routines should not be called with a lock held. Now I'm not sure this fixes the weird behavior I've seen, but can't hurt to still fix this buggy call.

eaburns committed:
Offer mime type UTF8_STRING for snarf.

For whatever reason when offering only text/plain, the snarf can't paste into Firefox. Offering also UTF8_STRING seems to fix it.

eaburns committed:
Disable debug by default.

eaburns committed:
wayland: abort compose when the keyboard leaves.

This indicates window focus change. Without this, focus change caused by alt-key combinations (ctrl+alt+arrow on many Linux configurations will switch virtual desktops) may incorrectly be interpreted as a key-compose.

lobre commented:
Hey, I have thought a lot about this Wayland backend implementation, trying also to learn about the Wayland world and get a better understanding of how a client should ideally behave.

I am still not 100% sure about everything, but I think the implementation could be laid out a bit differently for a more robust concurrency model.

I think the main thing we could move toward is a single Wayland thread that owns the `wl_display` and all `wl_*` objects, with RPC threads just pushing commands to it. libwayland isn't apparently happy with concurrent access, and right now we still have `wl_display_dispatch()` running while other threads make `wl_*` calls under a lock. A queue + wake fd approach would keep all Wayland interaction serialized and avoid a whole class of races, without really making the code more complex.

Along the same lines, a poll‑based loop on that thread could be more robust. Using `wl_display_prepare_read/read_events` and polling on the Wayland fd plus a wake fd (and clipboard fds when needed) would avoid doing `roundtrip()` or blocking reads inside callbacks. That makes it easier to reason about the event loop and reduces the chance of stalling the compositor.

If input handling and rendering commits also stay on that one thread, we end up with one place where ordering and state changes happen. It matches the "single graphics thread" idea the code already relies on, and it should make behavior more predictable across different compositors.

It would be great to have real Wayland-wizards recommending on whether this is the way of architecturing this Wayland implementation. Just to be sure it is robust enough and "by the book".

If anyone is interested, I have tried to gather everything I read and my set of ideas for how a client implementation would ideally be laid out in [this Gist](https://gist.github.com/lobre/821756100533c9cdea9dd50e81603378), which I have lazily let the LLM format for easy reading.
